File: src/main/java/com/conference/website/integration/RedisMetricsClient.java
package com.conference.website.integration;

import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.time.Duration;

@Component
public class RedisMetricsClient implements MetricsClient {

    private static final Duration WRITE_LATENCY = Duration.ofMillis(40);
    private static final Duration READ_LATENCY = Duration.ofMillis(30);

    private final RedisMetricsDatabase metricsDatabase;

    public RedisMetricsClient(RedisMetricsDatabase metricsDatabase) {
        this.metricsDatabase = metricsDatabase;
    }

    @Override
    public Mono<Long> incrementViews(Long talkId) {
        return Mono.fromCallable(() -> metricsDatabase.incrementAndGetViews(talkId))
                .delayElement(WRITE_LATENCY);
    }

    @Override
    public Mono<Long> getViews(Long talkId) {
        return Mono.fromCallable(() -> metricsDatabase.getViews(talkId))
                .delayElement(READ_LATENCY);
    }
}

File: src/main/kotlin/com/conference/website/integration/RedisMetricsClient.kt
package com.conference.website.integration

import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import java.time.Duration

@Component
class RedisMetricsClient(
    private val metricsDatabase: RedisMetricsDatabase
) : MetricsClient {

    override fun incrementViews(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.incrementAndGetViews(talkId) }
            .delayElement(WRITE_LATENCY)

    override fun incrementLikes(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.incrementAndGetLikes(talkId) }
            .delayElement(WRITE_LATENCY)

    override fun incrementAttends(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.incrementAndGetAttends(talkId) }
            .delayElement(WRITE_LATENCY)

    override fun getViews(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.getViews(talkId) }
            .delayElement(READ_LATENCY)

    override fun getLikes(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.getLikes(talkId) }
            .delayElement(READ_LATENCY)

    override fun getAttends(talkId: Long): Mono<Long> =
        Mono.fromCallable { metricsDatabase.getAttends(talkId) }
            .delayElement(READ_LATENCY)

    companion object {
        private val WRITE_LATENCY: Duration = Duration.ofMillis(40)
        private val READ_LATENCY: Duration = Duration.ofMillis(30)
    }
}
